File: antia_numerical_methods@Beliavsky/c/fun/divdif0.c
/*	Interpolation using Newton's divided difference formula
	simplified version of DIVDIF without derivative calculation

	XB : (input) Value of x at which interpolation is required
	X : (input) Array of length NTAB containing x values
	F : (input) Array of length NTAB containing function values
		F[I] is the tabulated function value at X[I].
	NUSE : (input/output) Number of points to be used for interpolation
		After execution it will contain the number actually used
	NTAB : (input) Number of points in the table
	FB : (output) Array containing interpolated values
	    FB[I] should contain interpolation using I points
	    FB[NUSE] should be the final value
	AEPS : (input) Required accuracy
	IFLG : (input) Flag to decide whether nearest point has to be found
		If IFLG=0 find the nearest point to XB to start interpolation
		otherwise if IF1 is admissible use X[IF1] as the first point
	IF1 : (input/output) The first point to be used for interpolation
		when IFLG!=0
		If IFLG=0 then IF1 is set to the index of nearest point in X
		
	Error status is returned by the value of the function DIVDIF0.
		0 value implies successful execution
		21 implies NUSE<1, in which case it is set to MIN(6,NTAB)
		22 implies NUSE>NTAB or NMAX, in which case it is reduced
		23 implies interpolation has not converged to specified accuracy

	Required functions : NEARST
*/

#include <math.h>

int nearst(double xb, double x[], int ntab);

int divdif0(double xb, double x[], double f[], int *nuse, int ntab,
		double fb[], double aeps, int iflg, int *if1)


{
	int i,j,k,next,in,ip,nit,ier, nmax=10;
	double err,px,xn[11],xd[11];

/*	Find the nearest point */

	if(iflg == 0 || *if1 < 0 || *if1 >= ntab) {
		next=nearst(xb,x,ntab);
		*if1=next;
	}
	else next=*if1;

	fb[1]=f[next];
	xd[1]=f[next];
	xn[1]=x[next];
	ier=0;
	px=1.0;

/*	Points between IN and IP are used for interpolation */

	ip=next; in=next;

/*	Maximum number of points to be used for interpolation */
	nit=*nuse; if(nmax<nit) nit=nmax; if(ntab<nit) nit=ntab;
	if(*nuse>nmax || *nuse>ntab) ier=22;
	if(*nuse<1) {
		ier=21;
		nit=6; if(nmax<nit) nit=nmax; if(ntab<nit) nit=ntab;
	}
	*nuse=1;
		
/*	Calculate successive interpolation polynomial */
	for(j=2; j<=nit; ++j) {

/*	Choose the next nearest point to XB */
		if(in<=0 ) {
			ip=ip+1; next=ip;
		}
		else if(ip >= ntab-1) {
			in=in-1; next=in;
		}
		else if(fabs(xb-x[ip+1]) < fabs(xb-x[in-1]) ) {
			ip=ip+1; next=ip;
		}
		else {
			in=in-1; next=in;
		}

/*	Calculating the divided differences */
		xd[j]=f[next];
		xn[j]=x[next];
		for(k=j-1; k>=1; --k) xd[k]=(xd[k+1]-xd[k])/(xn[j]-xn[k]);

		px=px*(xb-xn[j-1]);
		err=xd[1]*px;
		fb[j]=fb[j-1]+err;
		*nuse=j;

		if(fabs(err) < aeps) return ier;
	}
	return 23;
}

File: projects/antia_numerical_methods@Beliavsky/f77/sub8/divdif0.f
!	Interpolation using Newton's divided difference formula
!	simplified version of DIVDIF without derivative calculation
!
!	XB : (input) value of x at which interpolation is required
!	X : (input) real array of length NTAB containing x values
!	F : (input) real array of length NTAB containing function values
!		F(I) is the tabulated function value at X(I).
!	NUSE : (input/output) Number of points to be used for interpolation
!		After execution it will contain the number actually used
!	NTAB : (input) Number of points in the table
!	FB : (output) Real array containing interpolated values
!	       FB(I) should contain interpolation using I points
!	       FB(NUSE) should be the final value
!	AEPS : (input) Required accuracy
!	IER : Error parameter, IER=0 if the execution is successful
!		IER=21 implies NUSE<1, in which case it is set to MIN(6,NTAB)
!		IER=22 implies NUSE>NTAB or NMAX, in which case it is reduced
!		IER=23 implies interpolation has not converged to specified accuracy
!	IFLG : (input) Flag to decide whether nearest point has to be found
!		If IFLG=0 find the nearest point to XB to start interpolation
!		otherwise if IF1 is admissible use X(IF1) as the first point
!	IF1 : (input/output) The first point to be used for interpolation
!		when IFLG.NE.0
!		If IFLG=0 then IF1 is set to the index of nearest point in X
!
!	Required routines : NEARST

      SUBROUTINE DIVDIF0(XB,X,F,NUSE,NTAB,FB,AEPS,IER,IFLG,IF1)
      IMPLICIT REAL*8(A-H,O-Z)
      PARAMETER(NMAX=10)
      DIMENSION X(NTAB),F(NTAB),FB(*),XN(NMAX),XD(NMAX)

!	Choose the first point
      IF(IFLG.EQ.0.OR.IF1.LT.1.OR.IF1.GT.NTAB) THEN
        NEXT=NEARST(XB,X,NTAB)
        IF1=NEXT
      ELSE
        NEXT=IF1
      ENDIF
      FB(1)=F(NEXT)
      XD(1)=F(NEXT)
      XN(1)=X(NEXT)
      IER=0
      PX=1.0

!	Points between IN and IP are used for interpolation
      IP=NEXT
      IN=NEXT

!	Maximum number of points to be used for interpolation
      NIT=MIN(NMAX,NUSE,NTAB)
      IF(NUSE.GT.NMAX.OR.NUSE.GT.NTAB) IER=22
      IF(NUSE.LT.1) THEN
        IER=21
        NIT=MIN(6,NTAB,NMAX)
      ENDIF
      NUSE=1

!	Calculate successive interpolation polynomial
      DO 5000 J=2,NIT

!	Choose the next nearest point to XB
        IF(IN.LE.1) GO TO 2200
        IF(IP.GE.NTAB) GO TO 2000
        IF(ABS(XB-X(IP+1)).LT.ABS(XB-X(IN-1))) GO TO 2200
2000    IN=IN-1
        NEXT=IN
        GO TO 2800
2200    IP=IP+1
        NEXT=IP

!	Calculating the divided differences
2800    XD(J)=F(NEXT)
        XN(J)=X(NEXT)
        DO 3000 K=J-1,1,-1
3000    XD(K)=(XD(K+1)-XD(K))/(XN(J)-XN(K))

        PX=PX*(XB-XN(J-1))
        ERR=XD(1)*PX
        FB(J)=FB(J-1)+ERR
        NUSE=J

        IF(ABS(ERR).LT.AEPS) RETURN
5000  CONTINUE

      IER=23
      END
